trin-core/src/utils/hash_delay_queue.rs

```diff
@@ -0,0 +1,395 @@
+        queue.insert_with_timeout(key.clone(), Duration::from_millis(5));
+
+        assert_pending!(poll!(queue));
+
+        sleep(Duration::from_millis(1)).await;
+
+        queue.reset_timeout(&key, Duration::from_millis(10));
+
+        assert_pending!(poll!(queue));
+
+        sleep(Duration::from_millis(5)).await;
+
+        assert_pending!(poll!(queue));
+
+        sleep(Duration::from_millis(5)).await;
+
+        let entry = assert_ready_ok_some!(poll!(queue));
+        assert_eq!(key, entry);
+
+        let entry = assert_ready!(poll!(queue));
+        assert!(entry.is_none());
+    }
+
+    #[tokio::test]
+    async fn multi_reset() {
+        time::pause();
+
+        let mut queue = task::spawn(HashDelayQueue::<String>::new(DELAY));
+        let key = "foo".to_string();
+
+        queue.insert_with_timeout(key.clone(), Duration::from_millis(50));
+
+        assert_pending!(poll!(queue));
+
+        queue.reset_timeout(&key, Duration::from_millis(10));
+
+        assert_pending!(poll!(queue));
+
+        sleep(Duration::from_millis(5)).await;
+
+        assert_pending!(poll!(queue));
+
+        queue.reset_timeout(&key, Duration::from_millis(10));
+
+        sleep(Duration::from_millis(5)).await;
+
+        assert_pending!(poll!(queue));
+
+        sleep(Duration::from_millis(5)).await;
+
+        let entry = assert_ready_ok_some!(poll!(queue));
+        assert_eq!(key, entry);
+
+        let entry = assert_ready!(poll!(queue));
+        assert!(entry.is_none());
+    }
+
+    #[tokio::test]
+    async fn reset_expired_entry() {
+        time::pause();
+
+        let mut queue = task::spawn(HashDelayQueue::<String>::new(DELAY));
+        let key = "foo".to_string();
+
+        queue.insert_with_timeout(key.clone(), Duration::ZERO);
+        queue.reset_timeout(&key, Duration::from_millis(10));
+
+        assert_pending!(poll!(queue));
+
+        sleep(Duration::from_millis(10)).await;
+
+        let entry = assert_ready_ok_some!(poll!(queue));
+        assert_eq!(key, entry);
+
+        let entry = assert_ready!(poll!(queue));
+        assert!(entry.is_none());
+    }
+
+    #[tokio::test]
+    async fn remove_existing() {
+        time::pause();
+
+        let mut queue = task::spawn(HashDelayQueue::<String>::new(DELAY));
+        let key = "foo".to_string();
+
+        queue.insert_with_timeout(key.clone(), Duration::from_millis(1));
+
+        assert_pending!(poll!(queue));
+
+        let removed = queue.remove(&key);
+        assert!(removed);
+
+        sleep(Duration::from_millis(5)).await;
+
+        let entry = assert_ready!(poll!(queue));
+        assert!(entry.is_none());
+    }
+
+    #[tokio::test]
+    async fn remove_nonexisting() {
+        time::pause();
+
+        let mut queue = task::spawn(HashDelayQueue::<String>::new(DELAY));
+        let key = "foo".to_string();
+
+        let removed = queue.remove(&key);
+        assert!(!removed);
+    }
+
+    #[tokio::test]
+    async fn clear() {
+        time::pause();
+
```

Comment:
nitpick - but it seems trivial to go ahead and bump up the # of entries here and in `len()` to something like 3. I wouldn't call it necessary, you just never know what bugs are hiding above the `0-1` range